Prepare repository for VAST v1.0.0-rc1 (#2035)
tenzir/vast: VERSIONING.md

# Versioning

> This document aims to provide an overview of how VAST is versioned, and what
> significance a version bump carries.

Reviewer: ```suggestion
This document provides an overview of versioning in VAST, covers the semantics of version bumps, and describes the individual components subject to versioning.
```

> 

Reviewer: Here, I'm missing a quick overview about the various pieces that we are versioning in VAST.

Run the `vast version` command to get VAST's current version, and the versions

> of its major dependencies that it currently uses:

Reviewer: This should go in a separate (sub)section about identifying the version.

```json
{
  "VAST": "v1.0.4-61-geed8cb41a3",
  "VAST Build Tree Hash": "0eb048d0eeda5be7081f0c3a659ff17f",
  "CAF": "0.17.6",
  "Apache Arrow": "6.0.1",
  "jemalloc": "5.2.1",
  "plugins": {
    "compaction": "v1.2.0-61-g9131346958",
    "matcher": "v2.1.4-23-gfb69e84ade",
  }
}
```

## Version Components

> VAST's version number consists of three mandatory and two optional parts:

Reviewer: The reader immediately wonders what the versioning scheme relates to most. We should mention upfront that it's SemVer'ish.

> `v<major>.<minor>.<patch>[-rc<candidate>][-<tag>]`. Every release on the default

Reviewer: ```suggestion
`v<major>.<minor>.<patch>[-rc<candidate>][-<tag>]`. Every release on the default
```

Have we ever discussed the `v` prefix? We currently don't use this, and I like it that way because it makes for easier parsing.

Author: It's a sorting thing: The CalVer releases would always show up as later releases when sorting reverse-lexicographically until we hit major version 2022. With the prefix, reverse-lexicographical sorting equals sorting by newest release date.

On a side note, GitHub recommends using a `v` prefix for release tags. They refused to let you create a tag that wasn't prefixed with `v` from the releases page up until a few months ago when they redesigned it.

Reviewer: > It's a sorting thing

Okay, I can see that.

> On a side note, GitHub recommends using a `v` prefix for release tags.

Interesting, can you provide a source?

Author: This is on the page for creating a new release:

<img width="347" alt="image" src="https://user-images.githubusercontent.com/4488655/149925985-143f9d08-4a4f-454e-a678-9a1e11c1ce4c.png">

branch has an associated annotated and gpg-signed tag of the same name.

### Major

> A major version increment indicates that the release included breaking changes.
> Such breaking changes include required changes to user configuration, removal of
> previously deprecated features, and updates to the minimum versions of required
> third-party dependencies.

Reviewer: ```suggestion
A major version increment indicates that the release includes breaking changes,
such as required changes to user configuration, removal of
previously deprecated features, and updates to the minimum versions of required
third-party dependencies.
```

> VAST's database format is guaranteed to be forward-compatible between
> consecutive major versions.

Reviewer: ```suggestion
VAST processes can connect to one another if the major and minor version match.
```

This seems the wrong place to point this out; I think a "Compatibility/Guarantees" section would be a better place.

### Minor

A minor version increment indicates changes to functionality, e.g., the addition
of new features and configuration, or the deprecation of existing features and
configuration.

> VAST processes can connect to one another if the major and minor version match.

Reviewer: Same comment as in above section.

### Patch

> A patch version increment indicates backwards-compatible bug fixes.

Author: I thought about this, but that would make the version a bit harder to parse. I'm open to changing this.

> A patch version increment indicates backwards-compatible bug fixes.

Reviewer: Does anything speak against just releasing `4.52.0`, i.e., have patch explicitly be 0? I think having variable-length versions with options in there are tricky to handle downstream.

From a workflow perspective, we will always decide whether cutting a patch release is important or whether it can wait until the next minor release. In other words, a patch release would be important enough to not wait and thereby indicates an urgent issue.

> A patch version increment indicates backwards-compatible bug fixes.

Author: 4.52.0 is what's currently in the document. It's simply a matter of preference whether you want to omit the patch version if it is 0.

> A patch version increment indicates backwards-compatible bug fixes.

Author: We talked about this in a quick meeting today, and have decided to omit the patch version when talking about releases publicly. I will mention that upfront in the document.

> ### Candidate (optional)

Reviewer: ```suggestion
### Release Candidate (optional)
```

Author: The section name exactly matches the format described in the h2 section. Renaming it here doesn't make sense.

> For release candidates, the tag begings with an additional `-rc<number>`, e.g.,
> `v1.2.3-rc1` indicates the first release candidate for the release `v1.2.3`.

Reviewer: ```suggestion
For release candidates, the tag ends with an additional `-rc<number>`, e.g.,
`v1.2.3-rc1` indicates the first release candidate for the release `v1.2.3`.
```

### Tag (optional)

> VAST's build system uses [`git-describe --abbrev=10 --long
> --dirty`][git-describe] for versioning. This includes a count of commits since
> the last release, the exact commit hash of the build abbreviated to 10
> characters, and an optional `-dirty` suffix if the build contained uncommited
> changes.

Reviewer: ```suggestion
VAST's build system uses the following [git][git-desribe] command to determine the version:

```
git-describe --abbrev=10 --long --dirty`
```

This includes a count of commits since the last release, the exact commit hash of the build abbreviated to 10 characters, and an optional `-dirty` suffix if the build contained uncommited changes.
```

Please also add example output.

The tag is omitted if `git-describe` is unavailable at build time, or the build

> runs outside of the VAST Git repository.

Reviewer: Please describe the implications and when this matters, from a use-case perspective.

[git-describe]: https://git-scm.com/docs/git-describe

## `libvast` API and ABI Versioning

The API and ABI of the C++ library `libvast` are considered unstable, i.e., VAST
v1.3.4 will function with `libvast` v1.3.4 only, and not with `libvast` v1.3.5.

> The shared object number of `libvast` increaes with every release to indicate

Reviewer: ```suggestion
The shared object number of `libvast` increases with every release to indicate
```

that every release is considered a breaking change to the ABI. It is calculated
as follows from the VAST version:

```

> SO number = 1'000'000 * major
>           +     1'000 * minor
>           +           * patch
> ```
> 
> E.g., the SO name for VAST `v1.23.45-67-g89abcdef12` is `libvast.so.1023045`.

Reviewer: ```suggestion
SO number = 1'000'000 * major
          +     1'000 * minor
          +           * patch
```

E.g., the SO name for VAST `v1.23.45-67-g89abcdef12` is `libvast.so.1023045`.
```

`SO` should be defined once; or even avoided (the less abbreviations, the better)

Author: I can just call it the shared object library name, I think, and avoid having to explain anything.

## Plugin Versioning

Plugin versions are optional. If unspecified, VAST's build system re-uses VAST's

> version number. Tenzir encourages plugin authors to version their plugins
> separately.

Reviewer: ```suggestion
version number. We encourage plugin authors to version their plugins
separately.
```

We should point out that plugin developers should ideally adhere to the same guidelines defined here.

> The version components exactly mirror the version components of VAST. The build
> system uses [`git-describe`] with the last commit to 

Reviewer: ```suggestion
The plugin version components exactly mirror the version components of VAST. The build
system uses [`git-describe`] with the last commit to 
```

Open-ended sentence...

> Note that due to `libvast`'s unstable ABI, dynamic VAST plugin libraries must
> link against the exact same version of `libvast` as the `vast` binary itself.

Reviewer: ```suggestion
Due to `libvast`'s unstable ABI, VAST plugins distributed as dynamic libraries must
link against the exact same version of `libvast` as the `vast` binary itself.
```

> ## Release Cadence

Reviewer: This should go into RELEASING.md, it's orthogonal to versioning.

Author: I think we can merge the two documents. I will add the content to here.

Author: Actually the whole content of RELEASING.md is irrelevant with this document. The CI checks are always enforced, and the name now needs to be decided upon based on the changes. Let's have one place less to maintain and have only this paragraph.

Reviewer: I still think it's the wrong document to describe the release procedures (here: cadence) in the document that's about versioning. What does @tenzir/backend think?

Author: One alternative would be to simply remove this section.

I would still want to delete RELEASING.md, as the document is both outdated and not adding any value.

Reviewer: Okay, let's remove it from this document and rethink it in the context of documentation on our releasing workflows/guidelines.

Author: Done.

runs outside of the VAST Git repository. For the above example, this would
result in `v1.23.45-rc1`, losing information that is relevant to developers
only.

[git-describe]: https://git-scm.com/docs/git-describe

## Compatibility and Guarantees

VAST's database format is guaranteed to be forward-compatible between
consecutive major versions, i.e., we guarantee that VAST `v2.*` works with a
database created by VAST `v1.*`, and if a conversion is required to do it on
startup. We do _not_ guarantee that such a conversion works when skipping major
versions. Upgrading from `v1.*` to `v3.*` may require running `v2.*` first.

VAST processes can connect to one another if the major and minor version match.

## `libvast` API and ABI Versioning

The API and ABI of the C++ library `libvast` are considered unstable, i.e., the
`vast` binary `v1.3.4` will function with `libvast` C++ library `v1.3.4` only,
and not with `v1.3.5`.

The shared object number of `libvast` increases with every release to indicate
that every release is considered a breaking change to the ABI. It is calculated
as follows from the VAST version:

```
SO number = 1'000'000 * major
          +     1'000 * minor

>           +           * patch

Reviewer 2: This does not match the agreed upon spec any more.
